runtime/src/bank.rs

```diff
@@ -906,18 +921,47 @@ impl Bank {
         lock_results: Vec<Result<()>>,
         max_age: usize,
         error_counters: &mut ErrorCounters,
-    ) -> Vec<Result<()>> {
+    ) -> Vec<Result<HashAgeKind>> {
         let hash_queue = self.blockhash_queue.read().unwrap();
         OrderedIterator::new(txs, iteration_order)
             .zip(lock_results.into_iter())
             .map(|(tx, lock_res)| {
-                if lock_res.is_ok()
-                    && !hash_queue.check_hash_age(&tx.message().recent_blockhash, max_age)
-                {
-                    error_counters.reserve_blockhash += 1;
-                    Err(TransactionError::BlockhashNotFound)
-                } else {
-                    lock_res
+                match lock_res {
+                    Ok(()) => {
+                        let message = tx.message();
```

Comment:
This code looks super fragile. You add about 4 levels of branching within a `map()` closure, and those ~2^4 code paths are tested with a single test.  How about breaking this up into small functions and testing all the edge cases with separate unit tests. Might need to add some helper functions to avoid copy-pasta for setup code.

Also, I'd be surprised to see all that array indexing is safe.

Reply:
Yeah. This isn't my favorite chunk.  I'll break up the test and see if I can't find some logical spots to split out this logic.  The indexing stuff should've been in the TODO.